src/messaging/ExchangeMgr.cpp

```diff
@@ -0,0 +1,344 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements the ExchangeManager class.
+ *
+ */
+
+#ifndef __STDC_FORMAT_MACROS
+#define __STDC_FORMAT_MACROS
+#endif
+
+#ifndef __STDC_LIMIT_MACROS
+#define __STDC_LIMIT_MACROS
+#endif
+
+#include <inttypes.h>
+#include <stddef.h>
+
+#include <core/CHIPCore.h>
+#include <core/CHIPEncoding.h>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+#include <support/CHIPFaultInjection.h>
+#include <support/CodeUtils.h>
+#include <support/RandUtils.h>
+#include <support/logging/CHIPLogging.h>
+
+using namespace chip::Encoding;
+using namespace chip::Inet;
+using namespace chip::System;
+
+namespace chip {
+
+/**
+ *  Constructor for the ExchangeManager class.
+ *  It sets the state to kState_NotInitialized.
+ *
+ *  @note
+ *    The class must be initialized via ExchangeManager::Init()
+ *    prior to use.
+ *
+ */
+ExchangeManager::ExchangeManager()
+{
+    mState = State::kState_NotInitialized;
+}
+
+CHIP_ERROR ExchangeManager::Init(SecureSessionMgrBase * sessionMgr)
+{
+    if (mState != State::kState_NotInitialized)
+        return CHIP_ERROR_INCORRECT_STATE;
+
+    mSessionMgr = sessionMgr;
+
+    mNextExchangeId = GetRandU16();
+
+    memset(ContextPool, 0, sizeof(ContextPool));
+    mContextsInUse = 0;
+
+    memset(UMHandlerPool, 0, sizeof(UMHandlerPool));
+    OnExchangeContextChanged = nullptr;
+
+    sessionMgr->SetDelegate(this);
+
+    mState = State::kState_Initialized;
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ExchangeManager::Shutdown()
+{
+    if (mSessionMgr != nullptr)
+    {
+        mSessionMgr = nullptr;
+    }
+
+    OnExchangeContextChanged = nullptr;
+
+    mState = State::kState_NotInitialized;
+
+    return CHIP_NO_ERROR;
+}
+
+ExchangeContext * ExchangeManager::NewContext(const uint64_t & peerNodeId, void * appState)
```

Comment:
Can we remove appState ? we can use ExchangeContextDelegate to replace it

Reply:
The ExchangeContext provides an opaque, NULL pointer (void *AppState) that can be used to store it state context, that allows trivial retrieval of that state on receipt of the response.

The context of this opaque value is totally up to the protocol, we should not mandate its value. For example, currently Echo protocol use this pointer as its state context and BDX protocol allocates a new 'DownloadContext' as its opaque value to track it's download as a logical whole.

When BDX protocol allocates a new EC, makes the 'opaque app state' pointer point to the 'DownloadContext' it just allocated, updates the 'OnMessageReceived' callback to point to a 'BDXMessageReceivedFunc', then sends the ReceiveInit message on that EC by called EC->SendMessage(...).